src/storage/checkpoints/checkpoint.cpp

```diff
@@ -0,0 +1,144 @@
+      return;
+    }
+    catalog::table_oid_t curr_table_oid = queue[0];
+    queue.erase(queue.begin());
+    queue_latch.unlock();
+    common::ManagedPointer<storage::SqlTable> curr_table = accessor->GetTable(curr_table_oid);
+
+    std::string out_file = GenFileName(db_oid, curr_table_oid);
+    std::ofstream f;
+
+    auto curr_data_table = curr_table->table_.data_table_;
+    const BlockLayout &layout = curr_data_table->GetBlockLayout();
+    const std::vector<col_id_t> varlens = layout.Varlens();
+    auto column_ids = layout.AllColumns();
+    size_t col_num = column_ids.size();
+    std::cout << "col num: " << col_num << std::endl;
+
+    curr_data_table->blocks_latch_.Lock();
+    const std::list<RawBlock *> blocks = curr_data_table->blocks_;
+    curr_data_table->blocks_latch_.Unlock();
+    f.open(path + out_file, std::ios::binary);
+
+    // record # of blocks in block list, # of varlen col and # of dict col in the table
+    if (f.is_open()) {
+      unsigned long block_num = blocks.size();
+      unsigned long var_col_num = varlens.size();
+      unsigned long dict_col_num = 0;
+      unsigned long insertion_head_index = std::distance(curr_data_table->blocks_.begin(), curr_data_table->insertion_head_);
+      std::cout << "block num: " << block_num << std::endl;
+      std::cout << "var col num: " << var_col_num << std::endl;
+      std::cout << "dict col num: " << dict_col_num << std::endl;
+
+      ArrowBlockMetadata &metadata = curr_data_table->accessor_.GetArrowBlockMetadata(blocks.front());
+      // count dict_col num
+      for (auto i = 0u; i < col_num; i++) {
+        col_id_t col_id = column_ids[i];
+        ArrowColumnInfo &col_info = metadata.GetColumnInfo(layout, col_id);
+        if (col_info.Type() == ArrowColumnType::DICTIONARY_COMPRESSED) {
+          dict_col_num += 1;
+        }
+      }
+      std::cout << "dict col num: " << dict_col_num <<std::endl;
+      f.write(reinterpret_cast<const char *>(&block_num), sizeof(unsigned long));
+      f.write(reinterpret_cast<const char *>(&var_col_num), sizeof(unsigned long));
+      f.write(reinterpret_cast<const char *>(&dict_col_num), sizeof(unsigned long));
+      f.write(reinterpret_cast<const char *>(&insertion_head_index), sizeof(unsigned long));
+    } else {
+      // failed to copy to disk, add the table_oid back to queue
+      queue_latch.lock();
+      queue.push_back(curr_table_oid);
+      queue_latch.unlock();
+      return;
+    }
+
+    // traverse each block in the block list
+    for (RawBlock *block : blocks) {
+      ArrowBlockMetadata &metadata = curr_data_table->accessor_.GetArrowBlockMetadata(block);
+
+      // first pass: record var len data
+      for (auto i = 0u; i < col_num; i++) {
+        col_id_t col_id = column_ids[i];
+        ArrowColumnInfo &col_info = metadata.GetColumnInfo(layout, col_id);
+        if (layout.IsVarlen(col_id) && col_info.Type() == ArrowColumnType::GATHERED_VARLEN) {
+          std::cout << "varchar" << std::endl;
```

Comment:
Comment above. Might want to check the entire code for print statements before final code drop.